go-controller/pkg/ovn/ovn.go

```diff
@@ -225,84 +258,187 @@ func getPodNamespacedName(pod *kapi.Pod) string {
+	metrics.MonitorIPSec(cm.nbClient)
+	if config.Metrics.EnableConfigDuration {
+		// with k=10,
+		//  for a cluster with 10 nodes, measurement of 1 in every 100 requests
+		//  for a cluster with 100 nodes, measurement of 1 in every 1000 requests
+		metrics.GetConfigDurationRecorder().Run(cm.nbClient, cm.kube, 10, time.Second*5, cm.defaultStopChan)
+	}
+
+	cm.podRecorder.Run(cm.sbClient, cm.defaultStopChan)
+
+	// Start and sync the watch factory to begin listening for events
+	if err := cm.watchFactory.Start(); err != nil {
+		return err
+	}
+	return nil
+}
+
 // NewOvnController creates a new OVN controller for creating logical network
 // infrastructure and policy
-func NewOvnController(ovnClient *util.OVNClientset, wf *factory.WatchFactory, stopChan <-chan struct{}, addressSetFactory addressset.AddressSetFactory,
-	libovsdbOvnNBClient libovsdbclient.Client, libovsdbOvnSBClient libovsdbclient.Client,
-	recorder record.EventRecorder) *Controller {
-	if addressSetFactory == nil {
-		addressSetFactory = addressset.NewOvnAddressSetFactory(libovsdbOvnNBClient)
-	}
-	svcController, svcFactory := newServiceController(ovnClient.KubeClient, libovsdbOvnNBClient, recorder)
-	egressSvcController := newEgressServiceController(ovnClient.KubeClient, libovsdbOvnNBClient, svcFactory, stopChan)
-	var hybridOverlaySubnetAllocator *subnetallocator.SubnetAllocator
-	if config.HybridOverlay.Enabled {
-		hybridOverlaySubnetAllocator = subnetallocator.NewSubnetAllocator()
-	}
-	return &Controller{
-		client: ovnClient.KubeClient,
-		kube: &kube.Kube{
-			KClient:              ovnClient.KubeClient,
-			EIPClient:            ovnClient.EgressIPClient,
-			EgressFirewallClient: ovnClient.EgressFirewallClient,
-			CloudNetworkClient:   ovnClient.CloudNetworkClient,
-		},
-		watchFactory:                 wf,
-		stopChan:                     stopChan,
-		masterSubnetAllocator:        subnetallocator.NewSubnetAllocator(),
-		hybridOverlaySubnetAllocator: hybridOverlaySubnetAllocator,
-		lsManager:                    lsm.NewLogicalSwitchManager(),
-		logicalPortCache:             newPortCache(stopChan),
-		namespaces:                   make(map[string]*namespaceInfo),
-		namespacesMutex:              sync.Mutex{},
-		externalGWCache:              make(map[ktypes.NamespacedName]*externalRouteInfo),
-		exGWCacheMutex:               sync.RWMutex{},
-		addressSetFactory:            addressSetFactory,
-		sharedNetpolPortGroups:       syncmap.NewSyncMap[*defaultDenyPortGroups](),
-		eIPC: egressIPController{
+func (cm *ControllerManager) NewOvnController(nadInfo *util.NetAttachDefInfo,
+	addressSetFactory addressset.AddressSetFactory) (*Controller, error) {
+	clusterIPNet, err := config.ParseClusterSubnetEntries(nadInfo.NetCidr)
+	if err != nil {
+		return nil, fmt.Errorf("cluster subnet %s for network %s is invalid: %v", nadInfo.NetCidr, nadInfo.NetName, err)
+	}
+
+	stopChan := cm.defaultStopChan
+	if nadInfo.IsSecondary {
+		stopChan = make(chan struct{})
+	}
```

Comment:
how are we stopping the other OvnController(s) of secondary networks?

Reply:
when the controller's net-attach-def get deleted.

Reply:
We also need to handle ovnkube-master shutdown.

Reply:
yes, that is protected by the defaultStopChan and defaultWg.

Reply:
I might be missing it but I don't see how the new stopChan and WaitGroup that are passed to secondary controllers are tied to the defaultStopChan and defaultWg.

Reply:
when the associated netw-attach-def CRD gets deleted, the stopChan of the second network will be stopped and wait on the waitgroup of that particular second network.

when ovnkube-master is shutdown, it needs to call delete handler of of each network, and then stop the defaultStopChan and wait on the default waitgroup.

Reply:
> when ovnkube-master is shutdown, it needs to call delete handler of of each network, and then stop the defaultStopChan and wait on the default waitgroup.

I see. I guess we are not seeing that in this PR.